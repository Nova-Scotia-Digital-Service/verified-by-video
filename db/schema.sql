BEGIN;
    CREATE TABLE sessions (
        id              uuid PRIMARY KEY DEFAULT gen_random_uuid(),
        created_at      timestamptz NOT NULL DEFAULT transaction_timestamp(),
        expires_at      timestamptz NOT NULL DEFAULT transaction_timestamp() + '10 minutes'::interval
    );
COMMIT;
