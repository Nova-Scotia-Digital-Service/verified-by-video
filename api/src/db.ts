import { Pool } from 'pg'

import config from './config'

export const pool = new Pool({
  user: config.PG_USER,
  password: config.PG_PASSWORD,
  host: config.PG_HOST,
  database: config.PG_DATABASE,
})
