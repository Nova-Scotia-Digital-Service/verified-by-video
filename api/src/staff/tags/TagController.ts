import * as TD from '../../types'

import { Get, Controller, UseGuards, Post, Body, Delete, Param } from '@nestjs/common'
import { ApiProperty, ApiBearerAuth as SwaggerRequireAuth } from '@nestjs/swagger'

import { AuthGuard } from '../auth/AuthGuard'

import { createTag, deleteTag, getTagList } from './TagData'

class TagBodySchema {
  @ApiProperty({ type: 'string' })
  text!: string
}

@Controller('/staff/tags')
export class TagController {
  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Get('/')
  public async getTagList(): Promise<TD.Tag[]> {
    return await getTagList()
  }

  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Post('/')
  public async createTag(@Body() { text }: TagBodySchema): Promise<TD.Tag> {
    return await createTag(text)
  }

  @UseGuards(AuthGuard)
  @SwaggerRequireAuth()
  @Delete('/:tag_id')
  public async deleteTag(@Param('tag_id') tag_id: string): Promise<void> {
    await deleteTag(tag_id)
  }
}
