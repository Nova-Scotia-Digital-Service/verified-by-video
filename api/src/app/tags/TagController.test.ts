import { pool } from '../../db'
import { unpopulateDb, populateDb } from '../../../test/testUtils'

import { TagController } from './TagController'

const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/

beforeAll(async () => {
  await populateDb()
})

afterAll(async () => {
  await unpopulateDb()
  await pool.end()
})

// replace variable uuids in snapshot with constant string
expect.addSnapshotSerializer({
  serialize: () => 'Any<UUID>',
  test: (val) => val && typeof val === 'string' && UUID_REGEX.test(val),
})

describe('TagController', () => {
  describe('getTagList', () => {
    it('returns a list of tags', async () => {
      expect(await new TagController().getTagList()).toMatchSnapshot()
    })
  })

  describe('createTag', () => {
    it('adds a new tag to list of tags', async () => {
      await new TagController().createTag({ text: 'Some new tag' })
      expect(await new TagController().getTagList()).toMatchSnapshot()
    })
  })

  describe('deleteTag', () => {
    it('deletes a tag from the list of tags', async () => {
      const tagToDelete = { id: '1529c4ad-c5f7-404c-b880-6ffc9ad4ca1c', text: 'Tag 1' }

      const beforeList = await new TagController().getTagList()
      expect(beforeList).toContainEqual(tagToDelete)
      await new TagController().deleteTag(tagToDelete.id)
      const afterList = await new TagController().getTagList()
      expect(afterList).not.toContainEqual(tagToDelete)
    })
  })
})
