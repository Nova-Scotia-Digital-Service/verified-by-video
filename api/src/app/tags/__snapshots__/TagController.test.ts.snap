// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagController createTag adds a new tag to list of tags 1`] = `
[
  {
    "id": Any<UUID>,
    "text": "Tag 1",
  },
  {
    "id": Any<UUID>,
    "text": "Tag 2",
  },
  {
    "id": Any<UUID>,
    "text": "Tag 3",
  },
  {
    "id": Any<UUID>,
    "text": "Some new tag",
  },
]
`;

exports[`TagController getTagList returns a list of tags 1`] = `
[
  {
    "id": Any<UUID>,
    "text": "Tag 1",
  },
  {
    "id": Any<UUID>,
    "text": "Tag 2",
  },
  {
    "id": Any<UUID>,
    "text": "Tag 3",
  },
]
`;
